### DISCUSSION TITLE
Making the translations fetched, granular and consistent

### DISCUSSION THREAD

**Discussion by gomesalexandre** [shapeshift/web#1063] [Ideas]
Currently, we only have translations in a big `messages-<xy>`.json file.

It was originally a single file, but we now have multiple files, one for each language. That brings a first question:

### 1. How do we want to keep the translations consistent?

As we add more translations in `messages-en.json`, we are accumulating missing translations in others. The values are obviously going to be missing until we get translations done, but we are also missing the keys.

IMO, the best way to handle this currently - if this is something we want - is to create a simple script that would use diff the english JSON and the others, and add missing entries as "TODO".

Now, another issue we have is that the "single" (declined in multiple languages) file is getting bigger and bigger, harder to maintain, and for contributors, it is very easy to add translations to a path that's actually wrong semantically for the current domain. Which brings me to a second issue:

### 2. How do we want to make translations more granular?

A naive approach would be to have smaller JSON by domains, and construct the root JSON out of this, à la `combineReducers()`. I'm sure there are better approaches, and @cjthompson had one that has plugins in mind.

Finally, and after agreeing on solutions for these two first points, we should take into consideration a final one:

@0xdef1cafe [mentioned](https://github.com/shapeshift/web/pull/1047#issuecomment-1049172607) that we might use a third-party service provider in the future for translations.

### 3. How can we make translations compliant with a third-party service provider?

Obviously, this is all dependent on the service used. I haven't yet dug into the existing options and didn't find any mention of this on oneDAO, so I'm leaving this open for discussion and ideation.

**Comment by cjthompson** [comment]
There are a few questions that I think we need to answer first before talking about the technical implementation questions.

## What are the goals regarding localization?
Examples could be:
1. Differentiate "language" from "locale"
2. Correct formatting of dates and currencies based on a locale
3. Translation based on language
4. User should be able to manually select a locale and/or language
5. The app should automatically select the locale from the browser
6. We can leverage locale for number/date formatting even if there isn't a language translation that matches the locale.
7. Fallback to English for all strings that don't exist for a given language
## What are our requirements for translations?"

Examples of possible requirements could be:
1. Plugins (`src/plugins`) are able to provide their own translation files
2. The build process should warn us when there is a discrepancy between the JSON structure of translations vs English (are they missing keys, or were keys renamed in English and not changed in other files).
3. Components should have their own translations vs. a single translation file

## Legacy
https://beta.shapeshift.com has an implementation of localization that we can look at as an example. I used to run a browser with `fr-FR` as the locale to test all the date/currency formatting and translations just to test it.

**Comment by LPX55** [comment]
gm,

We are currently looking into FOSS CL (continuous localization) solutions. One in particular we are looking at is Firefox's Pontoon platform: https://shapeshift-cl.herokuapp.com/ -- however, the platform is not 100% set on, and I'm constantly on the lookout for a platform that could meet our needs even better. 

Whatever direction the engineering ws chooses, we have full confidence that careful thought and reasoning was put in and will support it. 

Lastly, please keep us in the loop so that we can adjust accordingly :)

edit: Also, I am in full support of future proofing the platform with a i18n and l10n plugin specifically for React. At the moment airbnb's polyglot is doing its job, but I believe there are more [advanced plugins](https://phrase.com/blog/posts/react-i18n-best-libraries/) that would make your guys' job much easier as the ShapeShift platform grows into who knows what in the future.

**Reply by LPX55** [reply]
Until the direction of a CL flow is somewhat finalized and integrated, I have also voiced my support for a simple script that would check the english translation file for new strings and have them appended into all languages as well as a simple json lint check.

**Comment by cjthompson** [comment]
Related discussion: https://github.com/shapeshift/web/discussions/1114
